# Проект 0: "Github. Самый быстрый старт"


## Цель проекта

Цель этого проекта - выдумать и имплементировать алгоритм принятия решений, следуя которому, *Игрок* в среднем сможет победить в простой разговорной *Игре* на числа (правила которой указаны ниже) **быстрее чем за 30 ходов**.

Среднее количество *Попыток* (за 1000 *Игр*) и будет служить критерием того насколько хорош алгоритм: чем меньше *Попыток* - тем лучше алгоритм 👍⭐️.

Вот правила игры:


### Правила Игры

В эту *Игру* играют двое: *Вода* и *Игрок*.
Вот какие *Ходы* им разрешено делать в процессе одной *Игры* и вот в каком по порядке:

1. *Вода* - загадывает целое *СекретноеЧисло* в интервале [1; 100]. После этого, *Ход* переходит к *Игроку*.

2. *Игрок* - придумывает и называет вслух одно *ПредположенноеЧисло* из интервала [1;100] и *Вода* сразу честно оглашает ему один из трёх исходов, который подходит под ситуацию:
- *ПредположенноеЧисло* **МЕНЬШЕ** *СекретногоЧисла*!
- *ПредположенноеЧисло* **РАВНО** *СекретномуЧислу*! 👍 *Победа*!
- *ПредположенноеЧисло* **БОЛЬШЕ** *СекретногоЧисла*!
После этого оглашения, считается что *Игрок* сделал одну *Попытку*.

3. *Ход* 2 повторяется сколько угодно раз - либо до того момента как *Игрок* не *Победит* 👍, либо пока он не признает *Поражение* ☠️.

4. Если Игрок *Победил*, то участники игры считают количество *Попыток* - и чем оно меньше - тем большим поводом для *Гордости* 👍 *Игроку* оно служит :)


### Стратегия

*Игрок* может придерживаться какого-то алгоритма чтобы немного увеличить свои шансы на быструю победу!

В этом ноутбуке я использую функцию которая эмулирует один такой алгоритм поведения *Игрока* чтобы проверить этот алгоритм на жизнеспособность.


## Идея/имплементация


### Идея/разработка

Мы хотим победить **побыстрее**. Поэтому задав один вопрос хочется получить как можно полезных ответов. Экономичность движений - драгоценна! 
    
1. Если у нас три исхода - меньше, равно, больше, то самое полезное мы можем получить в случае:
    ```
    oXo
    123
    ```
    Если мы зададим вопрос про число X (=2) то вот что будет означать каждый из разрешённых ответов *Воды*:
    - МЕНЬШЕ будет означать что СекретноеЧисло = 1.
    - РАВНО будет означать что СекретноеЧисло = 2.
    - БОЛЬШЕ будет означать что СекретноеЧисло = 3.
    
    Убить трёх приц одним камнем. У меня не получается представить ничего более выгодного. Может есть варианты лучше - но найти их не удаётся. Ладно, цепляемся пока за эту идею и проверим куда нас это направление мысли приведёт.
    
    
2. Наращиваем сложность - берём два `oXo`. По середине вставляем между ними один единственный элемент:
    ```
    oooXooo
    1  4  7
    ```
    - МЕНЬШЕ будет означать что СекретноеЧисло - одно из 123 и можно
      воспользоваться алгоритмом предыдущей сложности чтобы его найти за
      один вопрос.
    - РАВНО будет означать что СекретноеЧисло = 4.
    - БОЛЬШЕ будет означать что СекретноеЧисло - одно из 567 и можно
      воспользоваться алгоритмом предыдущей сложности чтобы его найти за
      один вопрос.
    
    
3. Если экстраполировать это простейшее поведение дальше, то получается повторящийся алгоритмический рисунок:
    ```
    oXo
    123
    
    oooXooo
    1  4  7
    
    oooooooXooooooo
    1      8      15
    
    oooooooooooooooXooooooooooooooo
    1              16             31
    
    oooooooooooooooooooooooooooooooXooooooooooooooooooooooooooooooo
    1                              32                             63
    
    ...
    ```
    
    Т.е.\
    1: [1; 3]   X=2=2<sup>1</sup>\
    2: [1; 7]   X=4=2<sup>2</sup>\
    3: [1; 15]  X=8=2<sup>3</sup>\
    4: [1; 31]  X=16=2<sup>4</sup>\
    5: [1; 63]  X=32=2<sup>5</sup>\
    6: [1; 127] X=64=2<sup>6</sup> - Как раз нам подходит чтобы угадать число из интервала [1; 100].
    
    Отсюда алгоритм - начинаем игру с числа 2<sup>6</sup>.
    - МЕНЬШЕ тогда пробуем (2<sup>6</sup> - 2<sup>5</sup>)
    - РАВНО успех - ПОБЕДА
    - БОЛЬШЕ тогда пробуем (2<sup>6</sup> + 2<sup>5</sup>)
    
    Если мы ещё не достигли успеха - просто понижаем степень.
    
    Интервал [1; 127] выплёскивается за пределы [1; 100], поэтому вижу возможность:
    можно ещё сэкономить *Попытки* способом который не нарушает правила -
        - если наше *ПредположенноеЧисло* вылезает за пределы [1; 100], мы НЕ
          ОГЛАШАЕМ его вслух и смотрим: если *Предположенное* число меньше 1,
          считаем что результат от *Воды* был **БОЛЬШЕ**,
        - если *ПредположенноеЧисло* больше верхнего предела, считаем что результат
          от *Воды* был **МЕНЬШЕ**
    В любом из этих двух результатов, продолжим двигаться по алгоритму,
    пока *ПредположенноеЧисло* не попадёт наконец в интервал [1; 100].

    На всякий случай чтобы не впасть в рекурсию - ставим себе
    ограничение: если мы почему-то посчитали *ПредположенноеЧисло* 12 раз и
    до сих пор не угадали результат - то надо *Сдаться*.
    
    Имея некоторый юмор, назовём счётчик который будет служить отсчётом
    ходов до того момента как *Игрок* сдастся - *УровеньОтчаянья*.
    
    Назовём это алгоритм - **v4**. Просто чтобы у него было какое-то имя.


## Имлементация алгоритма v4

In [1]:
def game_core_v4(secret_number, show_path=False):
    """Вернуть статистику действий Игрока, играющему по алгоритму v4.
    
    Именованные аргументы:
    secret_number -- число которое загадал Вода
    show_path -- булева - показывать ли в результате путь из
                 ПредположенныхЧисел которые придумывал Игрок: эта
                 часть статистики очень полезна для дебага но не
                 особенно нужна в остальных случаях (default False)
    
    На выходе - словарь со статистикой Игрока за одну игру.
    Поля словаря:
        is_win -- булева - победил ли Игрок? False если Сдался.
        attempt_count -- число - сколько Попыток сделал Игрок
        path -- список чисел - (опциональное поле - оно не показывается
                если аргумент show_path равен False) - все
                ПредположенныеЧисла которые обдумывал Игрок; в некоторых
                случаях тут может быть больше чисел чем было сделано
                Попыток - потому что Игрок не все числа решился Огласить
                вслух.
    
    Алгоритм:
    =========
    Начинаем игру с числа 2**6.
    - МЕНЬШЕ тогда пробуем (2**6 - 2**5)
    - РАВНО успех - ПОБЕДА
    - БОЛЬШЕ тогда пробуем (2**6 + 2**5)
    
    Если мы ещё не достигли успеха - просто понижаем степень.
    
    Если наше ПредположенноеЧисло вылезает за пределы [1; 100], мы НЕ
    ОГЛАШАЕМ его вслух и смотрим:
    - Если Предположенное число меньше 1, то считаем что результат от
      Воды был БОЛЬШЕ.
    - Если ПредположенноеЧисло больше верхнего предела, считаем что
      результат от Воды был МЕНЬШЕ.
    В любом из этих двух случаев - продолжить двигаться по алгоритму,
    пока ПредположенноеЧисло не попадёт наконец в интервал [1; 100].
    
    На всякий случай чтобы не впасть в рекурсию - ставим себе
    ограничение если мы почему-то посчитали ПредположенноеЧисло 12 раз и
    до сих пор не угадали результат - то надо Сдаться.
    """
    
    limit_min = 1
    limit_max = 100
    attempt_count = 0
    path = [] # Тут будут храниться все ПредположенныеЧисла.
    
    
    def declare(guess_number):
        """Огласить догадку вслух, истратив Попытку Игрока."""
        nonlocal attempt_count
        
        attempt_count += 1
        
        if guess_number < secret_number:
            return 'less'
        elif guess_number == secret_number:
            return 'equal'
        else:
            return 'more'
    
    
    # С точки зрения цикла можно считать что это - начальное положение
    # алгоритма.
    is_win = False # Если True - Игрок Победил, если False - Проиграл.
    guess_number = 0
    declaration = 'less' # ПредположенноеЧисло ... чем СекретноеЧисло.
    power = 6
    despair_level = 0 # Если Отчаянье достигнет 12 - Игрок сдастся.
    while despair_level < 12:
        
        despair_level += 1
        
        if declaration == 'equal':
            is_win = True
            break
        if declaration == 'less':
            guess_number += 2**power
        if declaration == 'more':
            guess_number -= 2**power
        
        path.append(guess_number)
        
        power -= 1
        
        # Если мы выбились за пределы.
        if guess_number < limit_min:
            declaration = 'less'
            continue
        if guess_number > limit_max:
            declaration = 'more'
            continue
        
        declaration = declare(guess_number)
    
    result = {
        'is_win': is_win,
        'attempt_count': attempt_count,
    }
    
    if show_path:
        result['path'] = path # Опциональная статистика.
    
    return result


Тут можно убедиться что функция вообще работает:

In [2]:
game_core_v4(42)

{'is_win': True, 'attempt_count': 6}

Тут можно проверить что эта функция работает по плану в краевых и странных случаях:

In [3]:
# Тут можно проверить как алгоритм v4 реагирует на разные
# СекретныеЧисла - особенно на крайние случаи и плохой ввод.
for test_range in [range(-1, 3), range(55, 54), [5.7], range(96, 102)]:
    print()
    print()
    print(f'test_range: {test_range}')
    print('=-' * 13)
    for secret_number in test_range:
        result = game_core_v4(secret_number, show_path=True)
        print()
        print(f'secret_number: {secret_number}')
        print(f'result: {result}')




test_range: range(-1, 3)
=-=-=-=-=-=-=-=-=-=-=-=-=-

secret_number: -1
result: {'is_win': False, 'attempt_count': 7, 'path': [64, 32, 16, 8, 4, 2, 1, 0.5, 0.75, 0.875, 0.9375, 0.96875]}

secret_number: 0
result: {'is_win': False, 'attempt_count': 7, 'path': [64, 32, 16, 8, 4, 2, 1, 0.5, 0.75, 0.875, 0.9375, 0.96875]}

secret_number: 1
result: {'is_win': True, 'attempt_count': 7, 'path': [64, 32, 16, 8, 4, 2, 1]}

secret_number: 2
result: {'is_win': True, 'attempt_count': 6, 'path': [64, 32, 16, 8, 4, 2]}


test_range: range(55, 54)
=-=-=-=-=-=-=-=-=-=-=-=-=-


test_range: [5.7]
=-=-=-=-=-=-=-=-=-=-=-=-=-

secret_number: 5.7
result: {'is_win': False, 'attempt_count': 12, 'path': [64, 32, 16, 8, 4, 6, 5, 5.5, 5.75, 5.625, 5.6875, 5.71875]}


test_range: range(96, 102)
=-=-=-=-=-=-=-=-=-=-=-=-=-

secret_number: 96
result: {'is_win': True, 'attempt_count': 2, 'path': [64, 96]}

secret_number: 97
result: {'is_win': True, 'attempt_count': 5, 'path': [64, 96, 112, 104, 100, 98, 97]}

secret

## Инструмент для оценки алгоритма

In [4]:
import numpy as np


def score_game(game_core):
    """Вернуть результаты проверки алгоритма угадывания.
    
    Именованные аргументы:
    game_core -- функция которая сможет эмулировать решения Игрока
    
    Запускаем игру 1000 раз, чтобы узнать, как быстро игра угадывает
    число.
    """
    results = []
    
    # Фиксируем RANDOM SEED, чтобы эксперимент был воспроизводимым!
    np.random.seed(1)
    secret_numbers = np.random.randint(1,101, size=1000)
    
    for secret_number in secret_numbers:
        results.append(game_core(secret_number))
    
    win_results = [x for x in results if x['is_win']]
    loss_results = [x for x in results if not x['is_win']]
    
    win_ratio = round(
        len(win_results) / len(secret_numbers),
        ndigits=4
        )
    loss_ratio = round(
        len(loss_results) / len(secret_numbers),
        ndigits=4
        )
    
    win_scores = [x['attempt_count'] for x in win_results]
    mean_win_score = round(np.mean(win_scores), 2)
    
    print()
    print()
    print(f'Ваш алгоритм:')
    print(f'* угадывает число в среднем за {mean_win_score} попыток')
    print(f'* проигрывает в среднем в {loss_ratio*100}% случаев')
    
    return {
        'win_ratio': win_ratio,
        'loss_ratio': loss_ratio,
        'mean_win_score': mean_win_score,
        }


Оцениваем алгоритм:

In [5]:
score_game(game_core_v4)



Ваш алгоритм:
* угадывает число в среднем за 5.99 попыток
* проигрывает в среднем в 0.0% случаев


{'win_ratio': 1.0, 'loss_ratio': 0.0, 'mean_win_score': 5.99}

`угадывает число в среднем за 5.99 попыток` - это быстрее чем `в среднем за 30 попыток` с которым я соревнуюсь! Похоже что я достиг успеха в своей цели ⭐️